### ISSUE TITLE
Add a slope graph

### ISSUE THREAD

**Issue by yishai7** [IsraelHikingMap/Site#948]
## Feature
(Delete the non-relevant)

## What I expect to happen
Add a slope graph similar to what available on Oruxmaps, Google Eart or ridewithgps.com

## My environment
* Operating System and version: 
* Web Browser and version:

## Thing I think the developers should know, images, links etc.
When planing off-road bike trip this can be great assistance. It helpps to understand the difficulty or even if certain section is can be riden.

**Comment by HarelM** [comment]
Thanks for taking the time and opening this issue!
I've added a comment in #854 which talks about improvements to the statistics window.
Closing this in favor of #854.

**Comment by HarelM** [comment]
Reopening since it will not be implemented as part of #854.
This is not as small implementation as I thought and will take some time to do it.
Since I personally do not see value for most users I'll be happy to receive a Pull-Request.

**Comment by yishai7** [comment]
Just to be on the same page - I have asked it for the website and not for the app.

**Comment by HarelM** [comment]
The website and the app mostly run the same code, I try to reduce the number of features that are specific to a client device.

**Comment by HarelM** [comment]
Let's see if the following is good enough:
![image](https://user-images.githubusercontent.com/3269297/81609899-4e5f1400-93e1-11ea-9f7c-07a53fafc110.png)
Will be available in next release.

**Comment by valleyofdawn** [comment]
Nice work.  I personally have rarely used this feature in other apps. I think mostly road-bike riders are interested in it.
1) The Y axis title should be changed from the current "height in m"
2) The graph needs extensive smoothing to be useful. It's currently too noisy and shows artifacts whenever the GPX vertical reading  jumps.
3) For smoothing' we can use moving average or moving median over a ~50m section. If you like I can play with height data in excel and suggest possible parameters.

Attaching a particularly bad example.
![image](https://user-images.githubusercontent.com/689201/81664734-7170de80-9448-11ea-8485-d92d8c9a532e.png)

**Comment by HarelM** [comment]
Slope values used in the chart are the same values that are presented in the box previous to this change. Feel free to suggest a formula to calculate them, I personally don't see a lot of value in complicating it, especially since no one mentioned there's a problem with it...
After playing a bit with it I think the following is better - using the right y axis to present the slope.
![image](https://user-images.githubusercontent.com/3269297/81667815-70da4700-944c-11ea-82df-cc14bc4e8988.png)
I'd like to see if anyone is using it first, otherwise I think it might be a good candidate to be removed.

**Comment by valleyofdawn** [comment]
There are more artifacts in GPS recording than in planned routes.

**Comment by HarelM** [comment]
I don't think the height graph is interesting for GPS recording in general as the height accuracy is just bad. Same goes to slope in this case...

**Comment by valleyofdawn** [comment]
As the saying goes, if it's worth doing, it's worth doing well. If you are happy with the result than leave it be. 
Height accuracy is good for [smartphones that have a barometer ](https://www.techrankup.com/en/smartphones-with-barometer/).
We could rely on DTM instead to draw the height data for other smartphones, but that's another issue.

**Comment by valleyofdawn** [comment]
With smoothing we can go from this to that.
![image](https://user-images.githubusercontent.com/689201/81675611-c2d49a00-9457-11ea-8d24-ccba38017a0d.png)

**Comment by valleyofdawn** [comment]
Regarding smoothing, please compare our result to that of another site you  have mentioned:
![image](https://user-images.githubusercontent.com/689201/82119807-1d3c6600-978a-11ea-81b3-263f7759608c.png)

**Comment by valleyofdawn** [comment]
The "zero slope" line they show is also useful.

**Comment by HarelM** [comment]
Added a moving mean which window is defined by the number of points in the chart - the more points the larger the window - window size is `number of points / 10`.
Will be published to site shortly, will be available in apps the next time we publish a version.
![image](https://user-images.githubusercontent.com/3269297/82138856-43b2dd80-982c-11ea-8d0f-a503343ef9d7.png)

**Comment by valleyofdawn** [comment]
Looks so much better! Thanks

**Comment by valleyofdawn** [comment]
The Y axis label on the right could be placed better.
![image](https://user-images.githubusercontent.com/689201/82139894-297cfd80-9834-11ea-987f-865f22303934.png)

**Comment by HarelM** [comment]
I preferred to keep the graph a bit bigger than placing this more to the right...

**Comment by yishai7** [comment]
הי,

 

כל הכבוד שאתם מיישמים זאת. מבחינת נראות בנוסף להערה של גיא זה יהיה נחמד שקו המגמה יהיה באותו צבע של הציר והשנתות הרלוונטים אליו. 

 

תודה

ישי

**Comment by valleyofdawn** [comment]
Yes. The graph size is important.
You can maybe gain a little space with an arrangement lie the one to the right here:
![image](https://user-images.githubusercontent.com/689201/82141010-af507700-983b-11ea-89cb-91c15778dfb9.png)

**Comment by HarelM** [comment]
Published to site, further changes are welcome using Pull Request. I'm not planning on changing it any further in relevance to this feature.

**Comment by HarelM** [comment]
@valleyofdawn @yishai7 Please check the site with the new slope graph - I have tried to get an impression of the slopes that are expected of a random route I did see here:

![image](https://user-images.githubusercontent.com/3269297/82143003-52f45400-9849-11ea-9b59-557324212056.png)
Before digging into the slope graph I thought that the above route has terrifying slopes, later on I found out that the slopes are minimal.
Should there be a minimal scale to slope of the chart to allow a user to instantly understand that a certain route is "easy" in terms of slopes?
For example instead of using a scale of -1% to 1% in the above chart using a scale of -20% to 20% would make this line almost flat.
Please let me know if you think a minimal value should be set to the slope y axis. Currently the scale is using the max between abs(minimal-slope) and maximal-slope and making the chart symmetric around zero.
Basically these lines:
https://github.com/IsraelHikingMap/Site/blob/master/IsraelHiking.Web/sources/application/components/route-statistics.component.ts#L690L691

**Comment by valleyofdawn** [comment]
If I understood correctly, we now divide the whole course into 100 sections (based on waypoints, not distance)  and average the slope in each?
I think we may have gone too far in the averaging algorithm, or maybe introduced a bug in the calculation.
Consider this graph for instance.
https://israelhiking.osm.org.il/share/EfXjZPiel4
It misjudges pretty long sections.
Specifically, there are several 1km long climbs in this route that register with a negative slope.

**Comment by HarelM** [comment]
So I understand you would like to remove smoothing all altogether. No problem.

**Comment by valleyofdawn** [comment]
Nope. I want to do it right. I can play with the parameters if you explain what you did.

**Comment by HarelM** [comment]
From the linked commit:
"Smoothing the slope by using a moving mean with window of number of point / 10"

**Comment by valleyofdawn** [comment]
So this will yield drastically different results with a planned route (typically sparse points) and a recorded track (dense points)?

**Comment by zstadler** [comment]
The route provided by the Facebook report has some strange effects associated with the length of the route. It is visible when you try to minimize the example to the relevant section.

For example, I have
- Located and marked a point with problematic -3% slope calculation in a positive slope point
   ![image](https://user-images.githubusercontent.com/1304610/84742176-c9a37f00-afb8-11ea-9d03-066f2f643abe.png)
- Split the route about 4 Km later. I assume there are more than 10 sampling points between the marked point and the split point.
- The slope in that point changed to +1%   
   ![image](https://user-images.githubusercontent.com/1304610/84741578-f3a87180-afb7-11ea-93f2-a1b5bece8696.png)

It is possible that this problem is related to the selection of points for the moving average calculation.

To reproduce:
- Open [this share](https://israelhiking.osm.org.il/share/VO7NR6lBDQ)
- The slope at the "-3%" point would be "1%'
- Join the previously-split route to its original length
- The slope at the "-3%" point shows "-3%"
I have split the original track

**Comment by valleyofdawn** [comment]
The moving average calculates a n/10 window for averaging, so a 100-point route will have a 10-point average and a 10,000 point route will have a 1000-point window, which can span several kilometres.
I suggest we cap the averaging window at 30 points while still maintaining the n/10 so that short routes would look okay.

**Comment by valleyofdawn** [comment]
The calculation is done towards the end of this code:
https://github.com/IsraelHikingMap/Site/blob/6c60d47f9c85cc2f8c30beb65a8769fad136bfbd/IsraelHiking.Web/sources/application/services/route-statistics.service.ts

**Comment by zstadler** [comment]
The distance between points along a route is likely to be uneven, especially when the route uses the underlying OSM topology
I suggest the window will be defined by distance. Some number between 50 and 500 meters seems reasonable to me.

**Comment by HarelM** [comment]
Please explain, preferably with examples using both recorded and planned routes of various lengths, why capping to 30 points or 50 to 500 meters is ideal.
I have tested it when developing with 20 points window and the results were not smooth enough in most cases.

**Comment by valleyofdawn** [comment]
Theoretically distance would make more sense than point count. Imagine a spliced route which is part-recorded part-planned. I couldn't test this easily with excel.

**Comment by yishai7** [comment]
The same route that Guy sent with slope graph of Ride with GPS website:

**Comment by zstadler** [comment]
@yishai7 ,

The mail attachments were not added to the Github issue.
Please open the [link to your message](https://github.com/IsraelHikingMap/Site/issues/948#issuecomment-644669793) in a browser, edit it and paste/upload the attachment.

> 
> The same route that Guy sent with slope graph of Ride with GPS website:
>

**Comment by yishai7** [comment]
![slope](https://user-images.githubusercontent.com/47181177/84770967-3e8aaf00-afe1-11ea-827a-b1407e908079.png)

**Comment by zstadler** [comment]
@yishai7,

I'm not sure I understand in what way the above example can help us. It also has instances of incorrect negative grade/slope, as [reported in Facebook](https://www.facebook.com/groups/israelhikingmap/permalink/3037439082977931/). For example:

![image](https://user-images.githubusercontent.com/1304610/84778196-4d765f00-afeb-11ea-925c-bcb69f4ba95b.png)

**Comment by HarelM** [comment]
The following is what I believe is the best solution to this issue.
This is using a regression LOESS algorithm.
Can be played with here:
https://observablehq.com/@harrystevens/loess-regression
The value which I found to be optimal is around 0.03 - which doesn't smooth too much but still removes the major artifacts.
![image](https://user-images.githubusercontent.com/3269297/84785240-ef01ae80-aff3-11ea-8aad-fe6f8d6d10bc.png)
Of course it can't solve all the problems, and I have only changed the visible line so the tooltip data is accurate while the chart itself is smooth:
![image](https://user-images.githubusercontent.com/3269297/84786021-de056d00-aff4-11ea-8f71-bdc307e0c546.png)
This is how it looks without any smoothing:
![image](https://user-images.githubusercontent.com/3269297/84786352-4fddb680-aff5-11ea-9ab3-bc59c7851905.png)

**Comment by valleyofdawn** [comment]
Looking good!

**Comment by zstadler** [comment]
If we cannot get the slope graph to be consistent with the slope numbers in the tooltip, I think we should remove it.
The above example shows a -5% slope tooltip on a +2.5% graph point.

![image](https://user-images.githubusercontent.com/1304610/84801513-f2069a00-b007-11ea-84bf-a798fe7ddbe4.png)

**Comment by HarelM** [comment]
I do not agree with the above statement.

**Comment by valleyofdawn** [comment]
I'm ok with them not being identical
